pkg/rpcbackend/wsbackend.go

```diff
@@ -0,0 +1,190 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcbackend
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/hyperledger/firefly-common/pkg/fftypes"
+	"github.com/hyperledger/firefly-common/pkg/i18n"
+	"github.com/hyperledger/firefly-common/pkg/log"
+	"github.com/hyperledger/firefly-common/pkg/wsclient"
+	"github.com/hyperledger/firefly-signer/internal/signermsgs"
+	"github.com/sirupsen/logrus"
+)
+
+// WSBackend performs communication with a backend
+type WSBackend interface {
+	CallRPC(ctx context.Context, method string, params ...interface{}) (id string, rpcErr *RPCError)
+	Subscribe(ctx context.Context, subChannel chan *RPCSubscriptionRequest, params ...interface{}) (error *RPCError)
+	UnsubscribeAll(ctx context.Context) (error *RPCError)
+	Connect(ctx context.Context) error
+	Close()
+}
+
+// NewRPCClient Constructor
+func NewWSRPCClient(client wsclient.WSClient) WSBackend {
+	return &WSRPCClient{
+		client:               client,
+		subscriptions:        make(map[string]chan *RPCSubscriptionRequest),
+		pendingSubscriptions: make(map[string]chan *RPCSubscriptionRequest),
+	}
+}
+
+type WSRPCClient struct {
+	client               wsclient.WSClient
+	requestCounter       int64
+	subscriptions        map[string]chan *RPCSubscriptionRequest
+	pendingSubscriptions map[string]chan *RPCSubscriptionRequest
+	pendingSubMutex      sync.Mutex
+	subMutex             sync.Mutex
+}
+
+func (rc *WSRPCClient) Connect(ctx context.Context) error {
+	if err := rc.client.Connect(); err != nil {
+		return err
+	}
+	go rc.receiveLoop(ctx)
+	return nil
+}
+
+func (rc *WSRPCClient) Close() {
+	rc.client.Close()
+}
+
+func (rc *WSRPCClient) allocateRequestID(req *RPCRequest) string {
+	reqID := fmt.Sprintf(`%.9d`, atomic.AddInt64(&rc.requestCounter, 1))
+	req.ID = fftypes.JSONAnyPtr(`"` + reqID + `"`)
+	return reqID
+}
+
+func (rc *WSRPCClient) Subscribe(ctx context.Context, subChannel chan *RPCSubscriptionRequest, params ...interface{}) (error *RPCError) {
+	rc.pendingSubMutex.Lock()
+	defer rc.pendingSubMutex.Unlock()
+	reqID, err := rc.CallRPC(ctx, "eth_subscribe", params...)
+	if err != nil {
+		return err
+	}
+	rc.pendingSubscriptions[reqID] = subChannel
+	return nil
+}
+
+func (rc *WSRPCClient) UnsubscribeAll(ctx context.Context) (error *RPCError) {
+	rc.subMutex.Lock()
+	for subID, subChan := range rc.subscriptions {
+		close(subChan)
+		delete(rc.subscriptions, subID)
+		_, err := rc.CallRPC(ctx, "eth_unsubscribe", subID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (rc *WSRPCClient) CallRPC(ctx context.Context, method string, params ...interface{}) (id string, rpcErr *RPCError) {
```

Comment:
This seems to be missing response handling... I'm worried about having a function in the common library that people start adopting, which we know is missing that fundamental utility.

If it's ok, I'm going to propose an extension to your PR to include response handling

Reply:
I've added that in